tests/components/ruckus_unleashed/test_device_tracker.py

```diff
@@ -0,0 +1,103 @@
+    DEFAULT_TITLE,
+    TEST_CLIENT,
+    TEST_CLIENT_ENTITY_ID,
+    init_integration,
+    mock_config_entry,
+)
+
+
+async def test_client_connected(hass):
+    """Test client connected."""
+    await init_integration(hass)
+
+    future = utcnow() + timedelta(minutes=60)
+    with patch(
+        "homeassistant.components.ruckus_unleashed.Ruckus.clients",
+        return_value={
+            TEST_CLIENT[CONF_MAC]: TEST_CLIENT,
+        },
+    ):
+        async_fire_time_changed(hass, future)
+        await hass.async_block_till_done()
+
+        await hass.helpers.entity_component.async_update_entity(TEST_CLIENT_ENTITY_ID)
+        test_client = hass.states.get(TEST_CLIENT_ENTITY_ID)
+        assert test_client.state == STATE_HOME
+
+
+async def test_client_disconnected(hass):
+    """Test client disconnected."""
+    await init_integration(hass)
+
+    future = utcnow() + timedelta(minutes=60)
+    with patch(
+        "homeassistant.components.ruckus_unleashed.Ruckus.clients",
+        return_value={},
+    ):
+        async_fire_time_changed(hass, future)
+        await hass.async_block_till_done()
+
+        await hass.helpers.entity_component.async_update_entity(TEST_CLIENT_ENTITY_ID)
+        test_client = hass.states.get(TEST_CLIENT_ENTITY_ID)
+        assert test_client.state == STATE_NOT_HOME
+
+
+async def test_clients_update_failed(hass):
+    """Test failed update."""
+    await init_integration(hass)
+
+    future = utcnow() + timedelta(minutes=60)
+    with patch(
+        "homeassistant.components.ruckus_unleashed.Ruckus.clients",
+        side_effect=ConnectionError,
+    ):
+        async_fire_time_changed(hass, future)
+        await hass.async_block_till_done()
+
+        await hass.helpers.entity_component.async_update_entity(TEST_CLIENT_ENTITY_ID)
+        test_client = hass.states.get(TEST_CLIENT_ENTITY_ID)
+        assert test_client.state == STATE_UNAVAILABLE
+
+
+async def test_restoring_clients(hass):
+    """Test restoring existing device_tracker entities if not detected on startup."""
+    entry = mock_config_entry()
+    entry.add_to_hass(hass)
+
+    registry = await entity_registry.async_get_registry(hass)
+    registry.async_get_or_create(
+        "device_tracker",
+        DOMAIN,
+        TEST_CLIENT[CONF_MAC],
+        suggested_object_id="ruckus_test_device",
+        config_entry=entry,
+    )
+
+    with patch(
+        "homeassistant.components.ruckus_unleashed.Ruckus.connect",
+        return_value=None,
+    ), patch(
+        "homeassistant.components.ruckus_unleashed.Ruckus.mesh_name",
+        return_value=DEFAULT_TITLE,
+    ), patch(
+        "homeassistant.components.ruckus_unleashed.Ruckus.clients",
+        return_value={},
+    ):
+        entry.add_to_hass(hass)
+        await hass.config_entries.async_setup(entry.entry_id)
+        await hass.async_block_till_done()
+
+    device = hass.states.get(TEST_CLIENT_ENTITY_ID)
+    assert device is not None
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6309587)